[LINUX KMS] Refresh rate switching does not work (debugged)

I'm trying kms modes and refresh rate switching does not work. for reasons I fail to understand.

I modified the code to output some useful info:
the file is: gfx/drivers_context/drm_ctx.c

i modified the snippet as follows:
```
static bool gfx_ctx_drm_set_video_mode(void *data,
      unsigned width, unsigned height,
      bool fullscreen)
{
```
[..]
```
   if ((width == 0 && height == 0) || !fullscreen)
   {
      if (width == 0 && height == 0)
         RARCH_WARN("[KMS]:w or h is zero\n");
      if (!fullscreen)
         RARCH_WARN("[KMS]:not fullscreen\n");      
      
      RARCH_WARN("[KMS]:koko w and h are %dx%d\n", width, height);
      RARCH_WARN("[KMS]:not fullscreen\n");       
      
      RARCH_WARN("[KMS]: Falling back to mode 0 (default)\n");
      g_drm_mode     = &g_drm_connector->modes[0];
   }
```
Then started a game at 50Hz and got that in the log:

```
[INFO] [Video]: Timings deviate too much. Will not adjust. (Target = 60.00 Hz, Game = 50.00 Hz)
[INFO] [Video]: Set video size to: 1536x864.
[WARN] [DRM]: Couldn't get device resources.
[WARN] [DRM]: Couldn't get device resources.
[WARN] [DRM]: Couldn't get device resources.
[INFO] [DRM]: Found 3 connectors.
[INFO] [DRM]: Connector 0 connected: no
[INFO] [DRM]: Connector 0 has 0 modes.
[INFO] [DRM]: Connector 1 connected: yes
[INFO] [DRM]: Connector 1 has 20 modes.
[INFO] [DRM]: Connector 1 assigned to monitor index: #1.
[INFO] [DRM]: Connector 2 connected: yes
[INFO] [DRM]: Connector 2 has 20 modes.
[INFO] [DRM]: Connector 2 assigned to monitor index: #2.
[INFO] [DRM]: Mode 0: (1920x1080) 1920 x 1080, 59.995499 Hz
[INFO] [DRM]: Mode 1: (1920x1080) 1920 x 1080, 75.002274 Hz
[INFO] [DRM]: Mode 2: (1920x1080) 1920 x 1080, 71.996422 Hz
[INFO] [DRM]: Mode 3: (1920x1080) 1920 x 1080, 49.995480 Hz
[INFO] [DRM]: Mode 4: (1680x1050) 1680 x 1050, 59.883251 Hz
[INFO] [DRM]: Mode 5: (1600x900) 1600 x 900, 60.000000 Hz
[INFO] [DRM]: Mode 6: (1280x1024) 1280 x 1024, 75.024673 Hz
[INFO] [DRM]: Mode 7: (1280x1024) 1280 x 1024, 60.019741 Hz
[INFO] [DRM]: Mode 8: (1152x864) 1152 x 864, 75.000000 Hz
[INFO] [DRM]: Mode 9: (1280x720) 1280 x 720, 60.000000 Hz
[INFO] [DRM]: Mode 10: (1280x720) 1280 x 720, 59.940201 Hz
[INFO] [DRM]: Mode 11: (1024x768) 1024 x 768, 75.028580 Hz
[INFO] [DRM]: Mode 12: (1024x768) 1024 x 768, 60.003841 Hz
[INFO] [DRM]: Mode 13: (832x624) 832 x 624, 74.551270 Hz
[INFO] [DRM]: Mode 14: (800x600) 800 x 600, 75.000000 Hz
[INFO] [DRM]: Mode 15: (800x600) 800 x 600, 60.316540 Hz
[INFO] [DRM]: Mode 16: (640x480) 640 x 480, 75.000000 Hz
[INFO] [DRM]: Mode 17: (640x480) 640 x 480, 60.000000 Hz
[INFO] [DRM]: Mode 18: (640x480) 640 x 480, 59.940475 Hz
[INFO] [DRM]: Mode 19: (720x400) 720 x 400, 70.081665 Hz
[INFO] [GLCore]: Found GL context: "kms".
[INFO] [GLCore]: Detecting screen resolution: 1920x1080.
[WARN] [KMS]:not fullscreen
[WARN] [KMS]:koko w and h are 1536x864
[WARN] [KMS]:not fullscreen
[WARN] [KMS]: Falling back to mode 0 (default)
```

Somehow instead of picking the nearest refresh rate, which appears to be Mode 3:
[INFO] [DRM]: Mode 3: (1920x1080) 1920 x 1080, 49.995480 Hz

It tries to syng to
[INFO] [DRM]: Mode 8: (1152x864) 1152 x 864, 75.000000 H

Then it detects it as not fullscreen.

Is the mode selection logic somehow flawed?

Uhm, i entered Xorg, switched to fullscreen, exited retroarch, went back to kms console and restarted retroarch there.

Refresh rate switching still wont work, but this time the log is clearer without warnings.

I'm kinda clueless.

```
[INFO] [Audio]: Set audio input rate to: 48000.00 Hz.
[INFO] [Video]: Set video size to: 1920x1080.
[ERROR] [Wayland]: Failed to connect to Wayland server.
[WARN] [DRM]: Couldn't get device resources.
[WARN] [DRM]: Couldn't get device resources.
[WARN] [DRM]: Couldn't get device resources.
[INFO] [DRM]: Found 3 connectors.
[INFO] [DRM]: Connector 0 connected: no
[INFO] [DRM]: Connector 0 has 0 modes.
[INFO] [DRM]: Connector 1 connected: yes
[INFO] [DRM]: Connector 1 has 20 modes.
[INFO] [DRM]: Connector 1 assigned to monitor index: #1.
[INFO] [DRM]: Connector 2 connected: yes
[INFO] [DRM]: Connector 2 has 20 modes.
[INFO] [DRM]: Connector 2 assigned to monitor index: #2.
[INFO] [DRM]: Mode 0: (1920x1080) 1920 x 1080, 59.995499 Hz
[INFO] [DRM]: Mode 1: (1920x1080) 1920 x 1080, 75.002274 Hz
[INFO] [DRM]: Mode 2: (1920x1080) 1920 x 1080, 71.996422 Hz
[INFO] [DRM]: Mode 3: (1920x1080) 1920 x 1080, 49.995480 Hz
[INFO] [DRM]: Mode 4: (1680x1050) 1680 x 1050, 59.883251 Hz
[INFO] [DRM]: Mode 5: (1600x900) 1600 x 900, 60.000000 Hz
[INFO] [DRM]: Mode 6: (1280x1024) 1280 x 1024, 75.024673 Hz
[INFO] [DRM]: Mode 7: (1280x1024) 1280 x 1024, 60.019741 Hz
[INFO] [DRM]: Mode 8: (1152x864) 1152 x 864, 75.000000 Hz
[INFO] [DRM]: Mode 9: (1280x720) 1280 x 720, 60.000000 Hz
[INFO] [DRM]: Mode 10: (1280x720) 1280 x 720, 59.940201 Hz
[INFO] [DRM]: Mode 11: (1024x768) 1024 x 768, 75.028580 Hz
[INFO] [DRM]: Mode 12: (1024x768) 1024 x 768, 60.003841 Hz
[INFO] [DRM]: Mode 13: (832x624) 832 x 624, 74.551270 Hz
[INFO] [DRM]: Mode 14: (800x600) 800 x 600, 75.000000 Hz
[INFO] [DRM]: Mode 15: (800x600) 800 x 600, 60.316540 Hz
[INFO] [DRM]: Mode 16: (640x480) 640 x 480, 75.000000 Hz
[INFO] [DRM]: Mode 17: (640x480) 640 x 480, 60.000000 Hz
[INFO] [DRM]: Mode 18: (640x480) 640 x 480, 59.940475 Hz
[INFO] [DRM]: Mode 19: (720x400) 720 x 400, 70.081665 Hz
[INFO] [GLCore]: Found GL context: "kms".
[INFO] [GLCore]: Detecting screen resolution: 1920x1080.
[INFO] [EGL] Found EGL client version >= 1.5, trying eglGetPlatformDisplay
[INFO] [EGL]: EGL version: 1.4
[INFO] [EGL]: Current context: 0x57001353d910.
[INFO] [KMS]: New FB: 1920x1080 (stride: 7680).
[INFO] [GLCore]: Vendor: Intel Open Source Technology Center, Renderer: Mesa DRI Intel(R) HD Graphics 4600 (HSW GT2).
[INFO] [GLCore]: Version: 4.5 (Core Profile) Mesa 21.3.9-arch.6 Amber.
[INFO] [GLCore]: Using resolution 1920x1080.
```

--- a/gfx/drivers/gl2.c
+++ b/gfx/drivers/gl2.c
@@ -4258,6 +4258,7 @@ static void *gl2_init(const video_info_t *video,
    unsigned temp_width                  = 0;
    unsigned temp_height                 = 0;
    bool force_smooth                    = false;
+   bool force_fullscreen                = false;
    const char *vendor                   = NULL;
    const char *renderer                 = NULL;
    const char *version                  = NULL;
@@ -4280,6 +4281,13 @@ static void *gl2_init(const video_info_t *video,
    if (gl->ctx_driver->get_video_size)
       gl->ctx_driver->get_video_size(gl->ctx_data,
                &mode_width, &mode_height);
+
+   if (!video->fullscreen && !gl->ctx_driver->has_windowed)
+   {
+      RARCH_DBG("[GL]: Config requires windowed mode, but context driver does not support it. "
+                "Forcing fullscreen for this session.\n");
+      force_fullscreen = true;
+   }
 
 #if defined(DINGUX)
    mode_width  = 320;
@@ -4311,17 +4319,23 @@ static void *gl2_init(const video_info_t *video,
       win_width  = full_x;
       win_height = full_y;
    }
+   /* If fullscreen had to be forced, video->width/height is incorrect */
+   else if (force_fullscreen)
+   {
+      win_width  = settings->uints.video_fullscreen_x;
+      win_height = settings->uints.video_fullscreen_y;
+   }
 
    if (     !gl->ctx_driver->set_video_mode
          || !gl->ctx_driver->set_video_mode(gl->ctx_data,
-            win_width, win_height, video->fullscreen))
+            win_width, win_height, (video->fullscreen || force_fullscreen)))
       goto error;
 #if defined(__APPLE__) && !defined(IOS) && !defined(HAVE_COCOA_METAL)
    /* This is a hack for now to work around a very annoying
     * issue that currently eludes us. */
    if (     !gl->ctx_driver->set_video_mode
          || !gl->ctx_driver->set_video_mode(gl->ctx_data,
-            win_width, win_height, video->fullscreen))
+            win_width, win_height, (video->fullscreen || force_fullscreen)))
       goto error;
 #endif
 
@@ -4446,7 +4460,7 @@ static void *gl2_init(const video_info_t *video,
    gl2_begin_debug(gl);
 #endif
 
-   if (video->fullscreen)
+   if (video->fullscreen || force_fullscreen)
       gl->flags  |=  GL2_FLAG_FULLSCREEN;
 
    mode_width     = 0;

--- a/gfx/drivers/gl3.c
+++ b/gfx/drivers/gl3.c
@@ -1770,6 +1770,7 @@ static void *gl3_init(const video_info_t *video,
    unsigned full_x, full_y;
    settings_t *settings                 = config_get_ptr();
    bool video_gpu_record                = settings->bools.video_gpu_record;
+   bool force_fullscreen                = false;
    int interval                         = 0;
    unsigned mode_width                  = 0;
    unsigned mode_height                 = 0;
@@ -1799,6 +1800,13 @@ static void *gl3_init(const video_info_t *video,
       gl->ctx_driver->get_video_size(gl->ctx_data,
                &mode_width, &mode_height);
 
+   if (!video->fullscreen && !gl->ctx_driver->has_windowed)
+   {
+      RARCH_DBG("[GLCore]: Config requires windowed mode, but context driver does not support it. "
+                "Forcing fullscreen for this session.\n");
+      force_fullscreen = true;
+   }
+
    full_x      = mode_width;
    full_y      = mode_height;
    mode_width  = 0;
@@ -1827,10 +1835,16 @@ static void *gl3_init(const video_info_t *video,
       win_width  = full_x;
       win_height = full_y;
    }
+   /* If fullscreen had to be forced, video->width/height is incorrect */
+   else if (force_fullscreen)
+   {
+      win_width  = settings->uints.video_fullscreen_x;
+      win_height = settings->uints.video_fullscreen_y;
+   }
 
    if (     !gl->ctx_driver->set_video_mode
          || !gl->ctx_driver->set_video_mode(gl->ctx_data,
-            win_width, win_height, video->fullscreen))
+            win_width, win_height, (video->fullscreen || force_fullscreen)))
       goto error;
 
    if (gl->flags & GL3_FLAG_USE_SHARED_CONTEXT)
@@ -1889,7 +1903,7 @@ static void *gl3_init(const video_info_t *video,
 
    if (video->vsync)
       gl->flags   |= GL3_FLAG_VSYNC;
-   if (video->fullscreen)
+   if (video->fullscreen || force_fullscreen)
       gl->flags   |= GL3_FLAG_FULLSCREEN;
    if (video->force_aspect)
       gl->flags   |= GL3_FLAG_KEEP_ASPECT;

--- a/gfx/drivers/vulkan.c
+++ b/gfx/drivers/vulkan.c
@@ -3318,6 +3318,7 @@ static void *vulkan_init(const video_info_t *video,
    int interval                       = 0;
    unsigned temp_width                = 0;
    unsigned temp_height               = 0;
+   bool force_fullscreen              = false;
    const gfx_ctx_driver_t *ctx_driver = NULL;
    settings_t *settings               = config_get_ptr();
 #ifdef VULKAN_HDR_SWAPCHAIN
@@ -3351,6 +3352,13 @@ static void *vulkan_init(const video_info_t *video,
       vk->ctx_driver->get_video_size(vk->ctx_data,
             &mode_width, &mode_height);
 
+   if (!video->fullscreen && !vk->ctx_driver->has_windowed)
+   {
+      RARCH_DBG("[Vulkan]: Config requires windowed mode, but context driver does not support it. "
+                "Forcing fullscreen for this session.\n");
+      force_fullscreen = true;
+   }
+
    full_x                             = mode_width;
    full_y                             = mode_height;
    mode_width                         = 0;
@@ -3376,10 +3384,16 @@ static void *vulkan_init(const video_info_t *video,
       win_width  = full_x;
       win_height = full_y;
    }
+   /* If fullscreen had to be forced, video->width/height is incorrect */
+   else if (force_fullscreen)
+   {
+      win_width  = settings->uints.video_fullscreen_x;
+      win_height = settings->uints.video_fullscreen_y;
+   }
 
    if (     !vk->ctx_driver->set_video_mode
          || !vk->ctx_driver->set_video_mode(vk->ctx_data,
-            win_width, win_height, video->fullscreen))
+            win_width, win_height, (video->fullscreen || force_fullscreen)))
    {
       RARCH_ERR("[Vulkan]: Failed to set video mode.\n");
       goto error;
@@ -3412,7 +3426,7 @@ static void *vulkan_init(const video_info_t *video,
       vk->flags         |=  VK_FLAG_VSYNC;
    else
       vk->flags         &= ~VK_FLAG_VSYNC;
-   if (video->fullscreen)
+   if (video->fullscreen || force_fullscreen)
       vk->flags         |=  VK_FLAG_FULLSCREEN;
    else
       vk->flags         &= ~VK_FLAG_FULLSCREEN;

--- a/gfx/video_driver.c
+++ b/gfx/video_driver.c
@@ -1243,7 +1243,8 @@ void video_switch_refresh_rate_maybe(
    unsigned video_bfi                 = settings->uints.video_black_frame_insertion;
    unsigned shader_subframes          = settings->uints.video_shader_subframes;
    bool vrr_runloop_enable            = settings->bools.vrr_runloop_enable;
-   bool exclusive_fullscreen          = settings->bools.video_fullscreen && !settings->bools.video_windowed_fullscreen;
+   bool exclusive_fullscreen          = (video_st->flags | VIDEO_FLAG_FORCE_FULLSCREEN) || (
+                                        settings->bools.video_fullscreen && !settings->bools.video_windowed_fullscreen);
    bool windowed_fullscreen           = settings->bools.video_fullscreen && settings->bools.video_windowed_fullscreen;
    bool all_fullscreen                = settings->bools.video_fullscreen || settings->bools.video_windowed_fullscreen;
 

